# Лабораторная работа №2
## по предмету "Системы искусственного интеллекта"
### Выполнила студентка группы БВТ2101 Пьянова Анна

В данной лабораторной работе вы будете работать с набором данных, который содержит информацию о технических характеристиках ноутбуков и их цену.
Целью работы является изучение теоретических основ методов машинного обучения.

В наборе данных для лабораторной работы содержится абор характеристик  мобильных телефонов, включая мощность аккумулятора, характеристики камеры, поддержку сети, память, размеры экрана и другие атрибуты. Столбец «price_range» классифицирует телефоны по ценовым диапазонам (этот столбец необходимо предсказать).

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

### Задание 1

Выгрузите данные из датасета. Изучите колонки, проверьте наличие пропусков. Постройте матрицу корреляции между признаками и целевой переменной. Сделайте выводы, что показывает эта матрица.

In [2]:
data = pd.read_csv('laptop.csv')
data.head()

battery_power  blue  clock_speed  dual_sim  fc  four_g  int_memory  m_dep  \
0            842     0          2.2         0   1       0           7    0.6   
1           1021     1          0.5         1   0       1          53    0.7   
2            563     1          0.5         1   2       1          41    0.9   
3            615     1          2.5         0   0       0          10    0.8   
4           1821     1          1.2         0  13       1          44    0.6   

   mobile_wt  n_cores  ...  px_height  px_width   ram  sc_h  sc_w  talk_time  \
0        188        2  ...         20       756  2549     9     7         19   
1        136        3  ...        905      1988  2631    17     3          7   
2        145        5  ...       1263      1716  2603    11     2          9   
3        131        6  ...       1216      1786  2769    16     8         11   
4        141        2  ...       1208      1212  1411     8     2         15   

   three_g  touch_screen  wifi  price_range  
0        0             0     1            1  
1        1             1     0            2  
2        1             1     0            2  
3        1             0     0            2  
4        1             1     0            1  

[5 rows x 21 columns]

In [3]:
data.isnull().sum()

battery_power    0
blue             0
clock_speed      0
dual_sim         0
fc               0
four_g           0
int_memory       0
m_dep            0
mobile_wt        0
n_cores          0
pc               0
px_height        0
px_width         0
ram              0
sc_h             0
sc_w             0
talk_time        0
three_g          0
touch_screen     0
wifi             0
price_range      0
dtype: int64

In [4]:
data.corr()

battery_power      blue  clock_speed  dual_sim        fc  \
battery_power       1.000000  0.011252     0.011482 -0.041847  0.033334   
blue                0.011252  1.000000     0.021419  0.035198  0.003593   
clock_speed         0.011482  0.021419     1.000000 -0.001315 -0.000434   
dual_sim           -0.041847  0.035198    -0.001315  1.000000 -0.029123   
fc                  0.033334  0.003593    -0.000434 -0.029123  1.000000   
four_g              0.015665  0.013443    -0.043073  0.003187 -0.016560   
int_memory         -0.004004  0.041177     0.006545 -0.015679 -0.029133   
m_dep               0.034085  0.004049    -0.014364 -0.022142 -0.001791   
mobile_wt           0.001844 -0.008605     0.012350 -0.008979  0.023618   
n_cores            -0.029727  0.036161    -0.005724 -0.024658 -0.013356   
pc                  0.031441 -0.009952    -0.005245 -0.017143  0.644595   
px_height           0.014901 -0.006872    -0.014523 -0.020875 -0.009990   
px_width           -0.008402 -0.041533    -0.009476  0.014291 -0.005176   
ram                -0.000653  0.026351     0.003443  0.041072  0.015099   
sc_h               -0.029959 -0.002952    -0.029078 -0.011949 -0.011014   
sc_w               -0.021421  0.000613    -0.007378 -0.016666 -0.012373   
talk_time           0.052510  0.013934    -0.011432 -0.039404 -0.006829   
three_g             0.011522 -0.030236    -0.046433 -0.014008  0.001793   
touch_screen       -0.010516  0.010061     0.019756 -0.017117 -0.014828   
wifi               -0.008343 -0.021863    -0.024471  0.022740  0.020085   
price_range         0.200723  0.020573    -0.006606  0.017444  0.021998   

                 four_g  int_memory     m_dep  mobile_wt   n_cores  ...  \
battery_power  0.015665   -0.004004  0.034085   0.001844 -0.029727  ...   
blue           0.013443    0.041177  0.004049  -0.008605  0.036161  ...   
clock_speed   -0.043073    0.006545 -0.014364   0.012350 -0.005724  ...   
dual_sim       0.003187   -0.015679 -0.022142  -0.008979 -0.024658  ...   
fc            -0.016560   -0.029133 -0.001791   0.023618 -0.013356  ...   
four_g         1.000000    0.008690 -0.001823  -0.016537 -0.029706  ...   
int_memory     0.008690    1.000000  0.006886  -0.034214 -0.028310  ...   
m_dep         -0.001823    0.006886  1.000000   0.021756 -0.003504  ...   
mobile_wt     -0.016537   -0.034214  0.021756   1.000000 -0.018989  ...   
n_cores       -0.029706   -0.028310 -0.003504  -0.018989  1.000000  ...   
pc            -0.005598   -0.033273  0.026282   0.018844 -0.001193  ...   
px_height     -0.019236    0.010441  0.025263   0.000939 -0.006872  ...   
px_width       0.007448   -0.008335  0.023566   0.000090  0.024480  ...   
ram            0.007313    0.032813 -0.009434  -0.002581  0.004868  ...   
sc_h           0.027166    0.037771 -0.025348  -0.033855 -0.000315  ...   
sc_w           0.037005    0.011731 -0.018388  -0.020761  0.025826  ...   
talk_time     -0.046628   -0.002790  0.017003   0.006209  0.013148  ...   
three_g        0.584246   -0.009366 -0.012065   0.001551 -0.014733  ...   
touch_screen   0.016758   -0.026999 -0.002638  -0.014368  0.023774  ...   
wifi          -0.017620    0.006993 -0.028353  -0.000409 -0.009964  ...   
price_range    0.014772    0.044435  0.000853  -0.030302  0.004399  ...   

               px_height  px_width       ram      sc_h      sc_w  talk_time  \
battery_power   0.014901 -0.008402 -0.000653 -0.029959 -0.021421   0.052510   
blue           -0.006872 -0.041533  0.026351 -0.002952  0.000613   0.013934   
clock_speed    -0.014523 -0.009476  0.003443 -0.029078 -0.007378  -0.011432   
dual_sim       -0.020875  0.014291  0.041072 -0.011949 -0.016666  -0.039404   
fc             -0.009990 -0.005176  0.015099 -0.011014 -0.012373  -0.006829   
four_g         -0.019236  0.007448  0.007313  0.027166  0.037005  -0.046628   
int_memory      0.010441 -0.008335  0.032813  0.037771  0.011731  -0.002790   
m_dep           0.025263  0.023566 -0.009434 -0.025348 -0.018388   0.017003   
mo

### Задание 2

Реализуйте с нуля алгоритм логистической регрессии для многоклассовой классификации.

In [5]:
X = data.drop('price_range', axis=1).values
y = data['price_range'].values

# Разделение данных на обучающую, валидационную и тестовую выборки
X_train, X_temp, y_train, y_temp = train_test_split(X, y, test_size=0.3, random_state=42)
X_val, X_test, y_val, y_test = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

# Нормализация данных
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_val = scaler.transform(X_val)
X_test = scaler.transform(X_test)

In [7]:
class LogisticRegression:
    def __init__(self, learning_rate=0.01, num_iterations=1000):
        self.learning_rate = learning_rate
        self.num_iterations = num_iterations
        self.weights = None
        self.bias = None

    def sigmoid(self, z):
        return 1 / (1 + np.exp(-z))

    def fit(self, X, y):
        num_samples, num_features = X.shape
        num_classes = len(np.unique(y))

        self.weights = np.zeros((num_classes, num_features))
        self.bias = np.zeros(num_classes)

        for i in range(num_classes):
            y_binary = np.where(y == i, 1, 0)

            for _ in range(self.num_iterations):
                linear_model = np.dot(X, self.weights[i]) + self.bias[i]
                y_predicted = self.sigmoid(linear_model)

                dw = (1 / num_samples) * np.dot(X.T, (y_predicted - y_binary))
                db = (1 / num_samples) * np.sum(y_predicted - y_binary)

                self.weights[i] -= self.learning_rate * dw
                self.bias[i] -= self.learning_rate * db

    def predict(self, X):
        linear_model = np.dot(X, self.weights.T) + self.bias
        y_predicted = self.sigmoid(linear_model)
        return np.argmax(y_predicted, axis=1)

    def predict_proba(self, X):
        linear_model = np.dot(X, self.weights.T) + self.bias
        y_predicted = self.sigmoid(linear_model)
        return y_predicted

    def accuracy(self, X, y):
        y_pred = self.predict(X)
        return np.mean(y_pred == y)

Обучите модель.

In [8]:
model = LogisticRegression(learning_rate=0.01, num_iterations=1000)
model.fit(X_train, y_train)

train_accuracy = model.accuracy(X_train, y_train)
val_accuracy = model.accuracy(X_val, y_val)
test_accuracy = model.accuracy(X_test, y_test)

print(f"Train Accuracy: {train_accuracy:.4f}")
print(f"Validation Accuracy: {val_accuracy:.4f}")
print(f"Test Accuracy: {test_accuracy:.4f}")

Train Accuracy: 0.6650
Validation Accuracy: 0.6567
Test Accuracy: 0.7033


### Задание 3

Реализуйте с нуля алгоритм Наивный Байес.

In [9]:
class NaiveBayes:
  def __init__(self):
      self.classes = None
      self.mean = None
      self.var = None
      self.priors = None

  def fit(self, X, y):
      self.classes = np.unique(y)
      num_classes = len(self.classes)
      num_features = X.shape[1]

      self.mean = np.zeros((num_classes, num_features), dtype=np.float64)
      self.var = np.zeros((num_classes, num_features), dtype=np.float64)
      self.priors = np.zeros(num_classes, dtype=np.float64)

      for idx, c in enumerate(self.classes):
          X_c = X[y == c]
          self.mean[idx, :] = X_c.mean(axis=0)
          self.var[idx, :] = X_c.var(axis=0)
          self.priors[idx] = X_c.shape[0] / float(X.shape[0])

  def _calculate_likelihood(self, mean, var, x):
      exponent = np.exp(- (x - mean) ** 2 / (2 * var))
      return exponent / np.sqrt(2 * np.pi * var)

  def _calculate_posterior(self, x):
      posteriors = []

      for idx, c in enumerate(self.classes):
          prior = np.log(self.priors[idx])
          class_conditional = np.sum(np.log(self._calculate_likelihood(self.mean[idx, :], self.var[idx, :], x)))
          posterior = prior + class_conditional
          posteriors.append(posterior)

      return self.classes[np.argmax(posteriors)]

  def predict(self, X):
      y_pred = [self._calculate_posterior(x) for x in X]
      return np.array(y_pred)

  def accuracy(self, X, y):
      y_pred = self.predict(X)
      return np.mean(y_pred == y)

Обучите модель.

In [10]:
model1 = NaiveBayes()
model1.fit(X_train, y_train)

train_accuracy1 = model1.accuracy(X_train, y_train)
val_accuracy1 = model1.accuracy(X_val, y_val)
test_accuracy1 = model1.accuracy(X_test, y_test)

print(f"Train Accuracy: {train_accuracy1:.4f}")
print(f"Validation Accuracy: {val_accuracy1:.4f}")
print(f"Test Accuracy: {test_accuracy1:.4f}")


Train Accuracy: 0.8157
Validation Accuracy: 0.7733
Test Accuracy: 0.8267


### Задание 4

Реализуйте с нуля алгоритм kNN.

In [29]:
class KNN:
    def __init__(self, k=3):
        self.k = k

    def fit(self, X_train, y_train):
        self.X_train = X_train
        self.y_train = y_train

    def euclidean_distance(self, x1, x2):
        return np.sqrt(np.sum((x1 - x2) ** 2))

    def predict(self, X_test):
        predictions = []
        for x in X_test:
            distances = [self.euclidean_distance(x, x_train) for x_train in self.X_train]
            k_indices = np.argsort(distances)[:self.k]
            k_nearest_labels = [self.y_train[i] for i in k_indices]
            predictions.append(self._most_common(k_nearest_labels))
        
        return np.array(predictions)

    def _most_common(self, labels):
        class_count = {}
        for label in labels:
            if label in class_count:
                class_count[label] += 1
            else:
                class_count[label] = 1
        
        most_common_label = max(class_count, key=class_count.get)
        return most_common_label

    def accuracy(self, X_test, y_test):
        predictions = self.predict(X_test)
        return np.sum(predictions == y_test) / len(y_test)

Обучите модель.

In [30]:
model2 = KNN()
model2.fit(X_train, y_train)

train_accuracy2 = model2.accuracy(X_train, y_train)
val_accuracy2 = model2.accuracy(X_val, y_val)
test_accuracy2 = model2.accuracy(X_test, y_test)

print(f"Train Accuracy: {train_accuracy2:.4f}")
print(f"Validation Accuracy: {val_accuracy2:.4f}")
print(f"Test Accuracy: {test_accuracy2:.4f}")

Train Accuracy: 0.8543
Validation Accuracy: 0.4867
Test Accuracy: 0.5033


### Задание 5

Вывод: в ходе выполнения лабораторной работы были обучены три различные модели. Из оценки точности можно сказать, что лучшие результаты были показаны алгоритмом Наивный Байес, так как высокая точность была достигнута не только на тренировочной, но и на валидационной и тестовой выборках. В свою очередь, алгоритм kNN показал точность на тренировочных данных = 85%, но лишь примерно 50% на тестовых и валидационных соответственно. Алгоритм же Логистическая Регрессия продемонтрировал не столь высокую точность на тренировочной выборке (67%) и примерно в том же диапазоне осталась точность на валидационной и тестовой выборках.